### PR TITLE
8278947: Support for array constants in constant table

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch adds support for arrays in compiled code constant tables and uses it for various vector replicate operations on x86.

Test: GHA, linux x64 tier 1-3

Thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278947](https://bugs.openjdk.java.net/browse/JDK-8278947): Support for array constants in constant table

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6908/head:pull/6908` \
`$ git checkout pull/6908`

Update a local copy of the PR: \
`$ git checkout pull/6908` \
`$ git pull https://git.openjdk.java.net/jdk pull/6908/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6908`

View PR using the GUI difftool: \
`$ git pr show -t 6908`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6908.diff">https://git.openjdk.java.net/jdk/pull/6908.diff</a>

</details>

### CODE REVIEW

**Comment by jatin-bhateja** [comment]
Hi @merykitty ,

This could be used to optimize some vector APIs like Vector.iota if all its inputs are constant by directly dumping the resultant constant array.

Another use could be to bring contents of the some constant generating stubs near to native method, but that may need detailed performance analysis since it may improve iCache hits but may increase over all code size which may impact the in-lining behavior.

Best Regards

**Review by vnkozlov — COMMENTED** [review]
Very nice.  Let me test it before approval.
You need second review.

**Review by vnkozlov — APPROVED** [review]
Testing passed.

**Comment by merykitty** [comment]
Hi, may anyone take a look at this PR, please. Thank you very much.

File: src/hotspot/share/opto/constantTable.cpp
Comment (DamonFool) [inline src/hotspot/share/opto/constantTable.cpp:44]: Why did you add `jint_cast`?

Does it mean the original impl for `T_FLOAD` wouldn't work as expected?

Comment (merykitty) [reply src/hotspot/share/opto/constantTable.cpp:44]: They are the same. The current one uses type punning to do a bit cast. The updated one just makes the cast more explicit and consolidates the theoretical UB of type punning so that we can fix it more easily in the future.
Thanks.

**Comment by merykitty** [comment]
@cl4es Yes I think we can utilize this to do so, we just need to expose a constant array and the load to the middle-end then.

File: src/hotspot/share/opto/constantTable.hpp
Comment (iwanowww) [inline src/hotspot/share/opto/constantTable.hpp:41]: Can you rely on `_type` and just use `T_ARRAY` tag for that purpose?

**Review by iwanowww — APPROVED** [review]
Nice improvement!

File: src/hotspot/share/opto/constantTable.hpp
Comment (merykitty) [reply src/hotspot/share/opto/constantTable.hpp:41]: In that case, an additional field would be needed to identify the type of the elements anyway.

**Comment by merykitty** [comment]
Thanks a lot for your testings and reviews.
/integrate

**Pull request closed without merging**
